README
NAME
    Device::SMBus - Perl interface for smbus using libi2c-dev library.

VERSION
    version 1.06

SYNOPSIS
       use Device::SMBus;
       $dev = Device::SMBus->new(
         I2CBusDevicePath => '/dev/i2c-1',
         I2CDeviceAddress => 0x1e,
       );
       print $dev->readByteData(0x20);

DESCRIPTION
    This is a perl interface to smbus interface using libi2c-dev library.

    Prerequisites:

    For Debian and derivative distros(including raspbian) use the following
    to install dependencies:

      sudo apt-get install libi2c-dev i2c-tools build-essential

    If you are using Angstrom Linux use the following:

      opkg install i2c-tools
      opkg install i2c-tools-dev

    For ArchLINUX use the following steps:

      pacman -S base-devel
      pacman -S i2c-tools

    Special Instructions for enabling the I2C driver on a Raspberry Pi:

    You will need to comment out the driver from the blacklist. currently
    the I2C driver isn't being loaded.

         sudo vim /etc/modprobe.d/raspi-blacklist.conf

    Replace this line

         blacklist i2c-bcm2708

    with this

         #blacklist i2c-bcm2708

    You now need to edit the modules conf file.

         sudo vim /etc/modules

    Add these two lines;

         i2c-dev
         i2c-bcm2708

    Now run this command(replace 1 with 0 for older model Pi)

         sudo i2cdetect -y 1

    If that doesnt work on your system you may alternatively use this:

         sudo i2cdetect -r 1

    you should now see the addresses of the i2c devices connected to your
    i2c bus

ATTRIBUTES
  I2CBusDevicePath
    Device path of the I2C Device.

     * On Raspberry Pi Model A this would usually be /dev/i2c-0 if you are using the default pins.
     * On Raspberry Pi Model B this would usually be /dev/i2c-1 if you are using the default pins.

  I2CDeviceAddress
    This is the Address of the device on the I2C bus, this is usually
    available in the device Datasheet.

     * for /dev/i2c-0 look at output of `sudo i2cdetect -y 0' 
     * for /dev/i2c-1 look at output of `sudo i2cdetect -y 1'

METHODS
  fileError
    returns IO::Handle->error() for the device handle since the last
    clearerr

  writeQuick
     $self->writeQuick($value)

    This sends a single bit to the device, at the place of the Rd/Wr bit.

  readByte
     $self->readByte()

    This reads a single byte from a device, without specifying a device
    register. Some devices are so simple that this interface is enough; for
    others, it is a shorthand if you want to read the same register as in
    the previous SMBus command

  writeByte
     $self->writeByte()

    This operation is the reverse of readByte: it sends a single byte to a
    device.

  readByteData
     $self->readByteData($register_address)

    This reads a single byte from a device, from a designated register. The
    register is specified through the Comm byte.

  writeByteData
     $self->writeByteData($register_address,$value)

    This writes a single byte to a device, to a designated register. The
    register is specified through the Comm byte. This is the opposite of the
    Read Byte operation.

  readNBytes
     $self->readNBytes($lowest_byte_address, $number_of_bytes);

    Read together N bytes of Data in linear register order. i.e. to read
    from 0x28,0x29,0x2a

     $self->readNBytes(0x28,3);

  readWordData
     $self->readWordData($register_address)

    This operation is very like Read Byte; again, data is read from a
    device, from a designated register that is specified through the Comm
    byte. But this time, the data is a complete word (16 bits).

  writeWordData
     $self->writeWordData($register_address,$value)

    This is the opposite of the Read Word operation. 16 bits of data is
    written to a device, to the designated register that is specified
    through the Comm byte.

  processCall
     $self->processCall($register_address,$value)

    This command selects a device register (through the Comm byte), sends 16
    bits of data to it, and reads 16 bits of data in return.

  writeBlockData
     $self->writeBlockData($register_address, $values)

    Writes a maximum of 32 bytes in a single block to the i2c device. The
    supplied $values should be an array ref containing the bytes to be
    written.

    The register address should be one that is at the beginning of a
    contiguous block of registers of equal length to the array of values
    passed. Not adhering to this will almost certainly result in unexpected
    behaviour in the device.

  DEMOLISH
    Destructor

CONSTANTS
  I2C_SLAVE
USAGE
    *   This module provides a simplified object oriented interface to the
        libi2c-dev library for accessing electronic peripherals connected on
        the I2C bus. It uses Moose.

SEE ALSO
    *   Moose

    *   IO::File

    *   Fcntl

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through github at
    <https://github.com/shantanubhadoria/device-smbus/issues>. You will be
    notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/shantanubhadoria/device-smbus>

      git clone git://github.com/shantanubhadoria/device-smbus.git

AUTHOR
    Shantanu Bhadoria <shantanu at cpan dott org>

CONTRIBUTORS
    *   Jonathan Stowe <jns+git@gellyfish.co.uk>

    *   Neil Bowers <neil@bowers.com>

    *   Shantanu Bhadoria <shantanu att cpan dott org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Shantanu Bhadoria.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Changelog for Device-SMBus

1.06 2013-12-30T05:17:47
 - spelling fixes [Shantanu Bhadoria]
 - Added more documentation, used mailmap to remove duplicates in
   contributors information, Thanks to git shortlog:
   https://www.kernel.org/pub/software/scm/git/docs/git-shortlog.html
   [Shantanu Bhadoria]
 - Merge pull request #2 from jonathanstowe/master
   
   add writeBlockData method [Shantanu Bhadoria]
 - add writeBlockData method [Jonathan Stowe]

1.05 2013-12-08T15:07:21
 - v1.05
   
    - adding spelling stopwords [Shantanu]
    - Merge branch 'master' of github.com:shantanubhadoria/Device-SMBus
      [Shantanu]
    - Adding information for Angstrom Linux [root] [Shantanu]
 - adding spelling stopwords [Shantanu]
 - Merge branch 'master' of github.com:shantanubhadoria/Device-SMBus
   [Shantanu]
 - Adding information for Angstrom Linux [root]

1.04 2013-12-08T15:04:49
 - v1.04
   
    - including i2c-dev.h with the distribution to Avoid issues arising
   from
      missing file and the munged up implementation each linux distro makes
      with this file [Shantanu] [Shantanu]
 - including i2c-dev.h with the distribution to Avoid issues arising from
   missing file and the munged up implementation each linux distro makes
   with this file [Shantanu]

1.03 2013-10-20T01:49:07
 - v1.03
   
    - Merge branch 'master' of github.com:shantanubhadoria/Device-SMBus
      [Shantanu]
    - Added more documentation [Shantanu] [Shantanu]
 - Merge branch 'master' of github.com:shantanubhadoria/Device-SMBus
   [Shantanu]
 - Added more documentation [Shantanu]

1.02 2013-10-16T05:13:47
 - v1.02
   
    - Fixed Changes File to meet spec [Shantanu Bhadoria] [Shantanu
   Bhadoria]
 - Fixed Changes File to meet spec [Shantanu Bhadoria]

1.01 2013-10-06T03:20:16
 - v1.01 [Shantanu]

1.00 2013-09-27T12:24:47
 - v0.11
   
    - Merge branch 'master' of github.com:shantanubhadoria/Device-SMBus
      [Shantanu]
    - Merge branch 'master' of github.com:shantanubhadoria/Device-SMBus
      [Shantanu Bhadoria]
    - minor changes [Shantanu Bhadoria] [Shantanu]
 - Merge branch 'master' of github.com:shantanubhadoria/Device-SMBus
   [Shantanu]
 - Merge branch 'master' of github.com:shantanubhadoria/Device-SMBus
   [Shantanu Bhadoria]
 - minor changes [Shantanu Bhadoria]

0.10 2013-09-27T12:23:43
 - v0.10
   
    - File Error catching logic [Shantanu] [Shantanu]
 - File Error catching logic [Shantanu]

0.09 2013-09-12T09:29:26
 - v0.09
   
    - Merge branch 'master' of github.com:shantanubhadoria/Device-SMBus
      [Shantanu]
    - Merge pull request #1 from neilbowers/master
   
      Reformatted Changes files as per CPAN::Changes::Spec [Shantanu
   Bhadoria] [Shantanu]
 - Merge branch 'master' of github.com:shantanubhadoria/Device-SMBus
   [Shantanu]
 - Merge pull request #1 from neilbowers/master
   
   Reformatted Changes files as per CPAN::Changes::Spec [Shantanu Bhadoria]

0.08 2013-09-12T09:27:16
 - v0.08
   
    - Merge pull request #1 from neilbowers/master
   
      Reformatted Changes files as per CPAN::Changes::Spec [Shantanu
   Bhadoria]
    - Reformatted as per CPAN::Changes::Spec [Neil Bowers] [Shantanu]
 - Merge pull request #1 from neilbowers/master
   
   Reformatted Changes files as per CPAN::Changes::Spec [Shantanu Bhadoria]
 - Reformatted as per CPAN::Changes::Spec [Neil Bowers]

0.07 2013-09-11T07:25:07
 - N Bytes [Shantanu]
 - N Byte [Shantanu Bhadoria]

0.06 2013-09-08T10:35:28
 - Switch from Moo to Moose to simplify my code [Shantanu Bhadoria]

0.05 2013-09-08T10:19:53
 - Added a few more easy compatibly functions [Shantanu Bhadoria]
 - Added remaining XS routines to the XS file [Shantanu Bhadoria]

0.04 2013-09-08T07:39:57
 - Final Check on  byteData methods [Shantanu Bhadoria]
 - Merge branch 'master' of github.com:shantanubhadoria/Device-SMBus
   
   Conflicts:
	   Changes
	   lib/Device/SMBus.pm [Shantanu Bhadoria]
 - Merge branch 'master' of github.com:shantanubhadoria/Device-SMBus
   
   Conflicts:
	   .gitignore
	   Changes
	   SMBus.xs
	   dist.ini
	   lib/Device/SMBus.pm [Shantanu Bhadoria]
 - Merge branch 'master' of github.com:shantanubhadoria/Device-SMBus
   
   Conflicts:
	   .gitignore
	   Changes
	   SMBus.xs
	   dist.ini
	   lib/Device/SMBus.pm [Shantanu Bhadoria]

0.03 2013-09-07T19:36:13
 - Merge branch 'master' of github.com:shantanubhadoria/Device-SMBus
   
   Conflicts:
	   .gitignore
	   Changes
	   SMBus.xs
	   dist.ini
	   lib/Device/SMBus.pm [Shantanu Bhadoria]
 - First Release with bare minumum function [Shantanu Bhadoria]

0.02 2013-09-07T19:29:21
 - First Release with bare minumum function [Shantanu Bhadoria]

0.01 2013-09-07T16:58:33
 - First Release with bare minumum function [Shantanu Bhadoria]



lib/Device/SMBus.pm
use strict;
use warnings;

package Device::SMBus;

# PODNAME: Device::SMBus
# ABSTRACT: Perl interface for smbus using libi2c-dev library.
#
# This file is part of Device-SMBus
#
# This software is copyright (c) 2013 by Shantanu Bhadoria.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
our $VERSION = '1.06'; # VERSION

# Dependencies
use 5.010000;

use Moose;
use Carp;

use IO::File;
use Fcntl;

require XSLoader;
XSLoader::load( 'Device::SMBus', $VERSION );


use constant I2C_SLAVE => 0x0703;


has I2CBusDevicePath => ( is => 'ro', );

has I2CBusFileHandle => (
    is      => 'ro',
    lazy    => 1,
    builder => '_build_I2CBusFileHandle',
);


has I2CDeviceAddress => ( is => 'ro', );

has I2CBusFilenumber => (
    is      => 'ro',
    lazy    => 1,
    builder => '_build_I2CBusFilenumber',
);

sub _build_I2CBusFileHandle {
    my ($self) = @_;
    my $fh = IO::File->new( $self->I2CBusDevicePath, O_RDWR );
    if ( !$fh ) {
        croak "Unable to open I2C Device File at $self->I2CBusDevicePath";
        return -1;
    }
    $fh->ioctl( I2C_SLAVE, $self->I2CDeviceAddress );
    return $fh;
}

# Implicitly Call the lazy builder for the file handle by using it and get the filenumber
sub _build_I2CBusFilenumber {
    my ($self) = @_;
    $self->I2CBusFileHandle->fileno();
}


sub fileError {
    my ($self) = @_;
    return $self->I2CBusFileHandle->error();
}


sub writeQuick {
    my ( $self, $value ) = @_;
    my $retval = Device::SMBus::_writeQuick( $self->I2CBusFilenumber, $value );
}


sub readByte {
    my ($self) = @_;
    my $retval = Device::SMBus::_readByte( $self->I2CBusFilenumber );
}


sub writeByte {
    my ( $self, $value ) = @_;
    my $retval = Device::SMBus::_writeByte( $self->I2CBusFilenumber, $value );
}


sub readByteData {
    my ( $self, $register_address ) = @_;
    my $retval = Device::SMBus::_readByteData( $self->I2CBusFilenumber,
        $register_address );
}


sub writeByteData {
    my ( $self, $register_address, $value ) = @_;
    my $retval = Device::SMBus::_writeByteData( $self->I2CBusFilenumber,
        $register_address, $value );
}


sub readNBytes {
    my ( $self, $reg, $numBytes ) = @_;
    my $retval = 0;
    $retval = ( $retval << 8 ) | $self->readByteData( $reg + $numBytes - $_ )
      for ( 1 .. $numBytes );
    return $retval;
}


sub readWordData {
    my ( $self, $register_address ) = @_;
    my $retval = Device::SMBus::_readWordData( $self->I2CBusFilenumber,
        $register_address );
}


sub writeWordData {
    my ( $self, $register_address, $value ) = @_;
    my $retval = Device::SMBus::_writeWordData( $self->I2CBusFilenumber,
        $register_address, $value );
}


sub processCall {
    my ( $self, $register_address, $value ) = @_;
    my $retval =
      Device::SMBus::_processCall( $self->I2CBusFilenumber, $register_address,
        $value );
}


sub writeBlockData {
    my ( $self, $register_address, $values ) = @_;

    my $value = pack "C*", @{$values};

    my $retval = Device::SMBus::_writeI2CBlockData( $self->I2CBusFilenumber,
        $register_address, $value );
    return $retval;
}

# Preloaded methods go here.


sub DEMOLISH {
    my ($self) = @_;
    $self->I2CBusFileHandle->close();
}

1;

__END__

=pod

=head1 NAME

Device::SMBus - Perl interface for smbus using libi2c-dev library.

=head1 VERSION

version 1.06

=head1 SYNOPSIS

   use Device::SMBus;
   $dev = Device::SMBus->new(
     I2CBusDevicePath => '/dev/i2c-1',
     I2CDeviceAddress => 0x1e,
   );
   print $dev->readByteData(0x20);

=head1 DESCRIPTION

This is a perl interface to smbus interface using libi2c-dev library. 

Prerequisites:

For Debian and derivative distros(including raspbian) use the following to install dependencies:

  sudo apt-get install libi2c-dev i2c-tools build-essential

If you are using Angstrom Linux use the following:

  opkg install i2c-tools
  opkg install i2c-tools-dev

For ArchLINUX use the following steps:

  pacman -S base-devel
  pacman -S i2c-tools

Special Instructions for enabling the I2C driver on a Raspberry Pi:

You will need to comment out the driver from the blacklist. currently the
I2C driver isn't being loaded.

     sudo vim /etc/modprobe.d/raspi-blacklist.conf

Replace this line 

     blacklist i2c-bcm2708

with this

     #blacklist i2c-bcm2708

You now need to edit the modules conf file.

     sudo vim /etc/modules

Add these two lines;

     i2c-dev
     i2c-bcm2708

Now run this command(replace 1 with 0 for older model Pi)

     sudo i2cdetect -y 1

If that doesnt work on your system you may alternatively use this:

     sudo i2cdetect -r 1

you should now see the addresses of the i2c devices connected to your i2c bus

=head1 ATTRIBUTES

=head2 I2CBusDevicePath

Device path of the I2C Device. 

 * On Raspberry Pi Model A this would usually be /dev/i2c-0 if you are using the default pins.
 * On Raspberry Pi Model B this would usually be /dev/i2c-1 if you are using the default pins.

=head2 I2CDeviceAddress

This is the Address of the device on the I2C bus, this is usually available in the device Datasheet.

 * for /dev/i2c-0 look at output of `sudo i2cdetect -y 0' 
 * for /dev/i2c-1 look at output of `sudo i2cdetect -y 1' 

=head1 METHODS

=head2 fileError

returns IO::Handle->error() for the device handle since the last clearerr

=head2 writeQuick

 $self->writeQuick($value)

This sends a single bit to the device, at the place of the Rd/Wr bit.

=head2 readByte

 $self->readByte()

This reads a single byte from a device, without specifying a device
register. Some devices are so simple that this interface is enough; for
others, it is a shorthand if you want to read the same register as in
the previous SMBus command

=head2 writeByte

 $self->writeByte()

This operation is the reverse of readByte: it sends a single byte
to a device. 

=head2 readByteData

 $self->readByteData($register_address)

This reads a single byte from a device, from a designated register.
The register is specified through the Comm byte.

=head2 writeByteData

 $self->writeByteData($register_address,$value)

This writes a single byte to a device, to a designated register. The
register is specified through the Comm byte. This is the opposite of
the Read Byte operation.

=head2 readNBytes

 $self->readNBytes($lowest_byte_address, $number_of_bytes);

Read together N bytes of Data in linear register order. i.e. to read from 0x28,0x29,0x2a 

 $self->readNBytes(0x28,3);

=head2 readWordData

 $self->readWordData($register_address)

This operation is very like Read Byte; again, data is read from a
device, from a designated register that is specified through the Comm
byte. But this time, the data is a complete word (16 bits).

=head2 writeWordData

 $self->writeWordData($register_address,$value)

This is the opposite of the Read Word operation. 16 bits
of data is written to a device, to the designated register that is
specified through the Comm byte.

=head2 processCall

 $self->processCall($register_address,$value)

This command selects a device register (through the Comm byte), sends
16 bits of data to it, and reads 16 bits of data in return.

=head2 writeBlockData

 $self->writeBlockData($register_address, $values)

Writes a maximum of 32 bytes in a single block to the i2c device.  The supplied $values should be
an array ref containing the bytes to be written.

The register address should be one that is at the beginning of a contiguous block of registers of equal length
to the array of values passed.  Not adhering to this will almost certainly result in unexpected behaviour in
the device.

=head2 DEMOLISH

Destructor

=head1 CONSTANTS

=head2 I2C_SLAVE

=head1 USAGE

=over

=item *

This module provides a simplified object oriented interface to the libi2c-dev library for accessing electronic peripherals connected on the I2C bus. It uses Moose.

=back

=head1 SEE ALSO

=over

=item *

L<Moose>

=item *

L<IO::File>

=item *

L<Fcntl>

=back

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through github at 
L<https://github.com/shantanubhadoria/device-smbus/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/shantanubhadoria/device-smbus>

  git clone git://github.com/shantanubhadoria/device-smbus.git

=head1 AUTHOR

Shantanu Bhadoria <shantanu at cpan dott org>

=head1 CONTRIBUTORS

=over 4

=item *

Jonathan Stowe <jns+git@gellyfish.co.uk>

=item *

Neil Bowers <neil@bowers.com>

=item *

Shantanu Bhadoria <shantanu att cpan dott org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Shantanu Bhadoria.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This Makefile.PL for  was generated by Dist::Zilla.
# Don't edit it but the dist.ini used to construct it.

use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Perl interface for smbus using libi2c-dev library.",
  "AUTHOR" => "Shantanu Bhadoria <shantanu at cpan dott org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DEFINE" => "",
  "DISTNAME" => "Device-SMBus",
  "EXE_FILES" => [],
  "INC" => "-I.",
  "LIBS" => [
    " "
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010000",
  "NAME" => "Device::SMBus",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Fcntl" => 0,
    "IO::File" => 0,
    "Moose" => 0,
    "XSLoader" => 0,
    "constant" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Find" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "List::Util" => 0,
    "Test::More" => 0
  },
  "VERSION" => "1.06",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



if  (eval {require ExtUtils::Constant; 1}) {
  # If you edit these definitions to change the constants used by this module,
  # you will need to use the generated const-c.inc and const-xs.inc
  # files to replace their "fallback" counterparts before distributing your
  # changes.
  my @names = (qw());
  ExtUtils::Constant::WriteConstants(
                                     NAME         => 'Device::SMbus',
                                     NAMES        => \@names,
                                     DEFAULT_TYPE => 'IV',
                                     C_FILE       => 'const-c.inc',
                                     XS_FILE      => 'const-xs.inc',
                                  );

}
else {
  use File::Copy;
  use File::Spec;
  foreach my $file ('const-c.inc', 'const-xs.inc') {
    my $fallback = File::Spec->catfile('fallback', $file);
    copy ($fallback, $file) or die "Can't copy $fallback to $file: $!";
  }
}
